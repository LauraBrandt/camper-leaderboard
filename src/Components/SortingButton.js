import React from 'react';
import PropTypes from 'prop-types';

const SortingButton = ({ type, direction, inactive, handleClick }) => {
  return(
    <div>
      <button 
        className={`${type} ${direction}${inactive ? ' inactive' : ''}`} 
        title={`Sort ${direction === 'up' ? 'descending' : 'ascending'}`}
        onClick={inactive ? null : handleClick}
      >
        <i className={`fas fa-caret-${direction}`}></i>
      </button>
    </div>
  );
}

SortingButton.propTypes = {
  type: PropTypes.string.isRequired,
  direction: PropTypes.oneOf(['up', 'down']),
  handleClick: PropTypes.func.isRequired,
  inactive: PropTypes.bool
}

export default SortingButton;
