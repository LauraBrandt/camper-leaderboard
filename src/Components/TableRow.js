import React from 'react';
import PropTypes from 'prop-types';

const TableRow = ({ user, index }) => {
  return(
    <tr className="row">
      <td>{index+1}</td>
      <td>
        <a href={`https://www.freecodecamp.org/${user.username}`} className="user-info">
          <img src={user.img} alt={`${user.username} avatar`}/>
          <p>{user.username}</p>
        </a>
      </td>
      <td>{user.recent}</td>
      <td>{user.alltime}</td>
    </tr>
  );
}

TableRow.propTypes = {
  index: PropTypes.number.isRequired,
  user: PropTypes.shape({
    username: PropTypes.string.isRequired,
    recent: PropTypes.number.isRequired,
    alltime: PropTypes.number.isRequired,
    img: PropTypes.string
  })
}

export default TableRow;
