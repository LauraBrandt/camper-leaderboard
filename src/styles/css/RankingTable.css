.ranking-table-wrapper {
  width: 100%;
  flex: 1 0 auto;
  padding: 3em 0; }

.ranking-table {
  background-color: white;
  width: 90%;
  margin: auto;
  border-radius: 7px;
  border-collapse: collapse;
  border-style: hidden;
  font-size: 1.1rem;
  font-family: 'Source Sans Pro', sans-serif; }

.message {
  text-align: center;
  font-size: 1.2rem;
  padding: 3.5em 1em; }

.header-row th {
  color: white;
  padding: 0.8em;
  background-color: #A4755F;
  font-weight: normal;
  font-size: 1.2rem; }
  .header-row th:first-child {
    border-top-left-radius: 7px; }
  .header-row th:last-child {
    border-top-right-radius: 7px; }

.row td {
  border: 2px solid rgba(164, 117, 95, 0.4);
  padding: 0.3em;
  text-align: center; }

.row:nth-child(even) {
  background-color: rgba(224, 224, 224, 0.5); }

.row:last-child td:first-child {
  border-bottom-left-radius: 7px; }

.row:last-child td:last-child {
  border-bottom-right-radius: 7px; }

.row:hover {
  background-color: rgba(164, 117, 95, 0.3); }

.user-info {
  text-decoration: none;
  color: #3949AB;
  display: flex;
  align-items: center; }
  .user-info img {
    width: 45px;
    height: 45px;
    border: 1px solid rgba(147, 147, 147, 0.4);
    border-radius: 3px;
    margin: 0.3em 1.5em 0.3em 1em; }
  .user-info p {
    margin: 0.5em 0; }

.points {
  display: flex;
  justify-content: space-between;
  align-items: center; }
  .points .points-sorters {
    display: flex;
    flex-direction: column; }
    .points .points-sorters button {
      padding: 0;
      margin: 0;
      border: none;
      background-color: #A4755F;
      color: #FFFFFF;
      font-size: 1rem;
      cursor: pointer; }
      .points .points-sorters button:hover, .points .points-sorters button:focus {
        color: #c7c7c7;
        outline: none; }
      .points .points-sorters button:active {
        color: #bababa; }
      .points .points-sorters button.inactive {
        display: none; }

@media (max-width: 450px) {
  .ranking-table-wrapper {
    padding: 0; }
  .ranking-table {
    width: 100%;
    font-size: 1rem;
    border-radius: 0; }
    .ranking-table .header-row th {
      font-size: 1rem;
      border-radius: 0; }
    .ranking-table .row td {
      border-radius: 0; }
  .user-info {
    flex-direction: column; }
    .user-info img {
      margin: 0.5em 0 0 0; } }
