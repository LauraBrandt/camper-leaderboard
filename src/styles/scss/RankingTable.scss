@import '_variables';

.ranking-table-wrapper {
  width: 100%;
  flex: 1 0 auto;
  padding: 3em 0;
}

.ranking-table {
  background-color: white;
  width: 90%;
  margin: auto;
  border-radius: $table-border-radius;
  border-collapse: collapse;
  border-style: hidden;
  font-size: 1.1rem;
  font-family: 'Source Sans Pro', sans-serif;
}


.header-row {
  th {
    color: white;
    padding: 0.8em;
    background-color: $fcc-brown;
    font-weight: normal;
    font-size: 1.2rem;
    &:first-child {
      border-top-left-radius: $table-border-radius;
    }
    &:last-child {
      border-top-right-radius: $table-border-radius;
    }
  }
}

.row {
  td {
    border: 2px solid rgba($fcc-brown, 0.4);
    padding: 0.3em;
    text-align: center;
  }
  &:nth-child(even) {
    background-color: rgba($fcc-grey, 0.5);
  }
  &:last-child {
    td:first-child {
      border-bottom-left-radius: $table-border-radius;
    }
    td:last-child {
      border-bottom-right-radius: $table-border-radius;
    }
  }
  &:hover {
    background-color: rgba($fcc-brown, 0.3);
  }
}

.user-info {
  text-decoration: none;
  color: $fcc-blue;
  display: flex;
  align-items: center;
  img {
    width: $image-size;
    height: $image-size;
    border: 1px solid rgba($fcc-dark-grey, 0.4);
    border-radius: 3px;
    margin: 0.3em 1.5em 0.3em 1em;
  }
  p {
    margin: 0.5em 0;
  }
}

@media (max-width: 450px) {
  .ranking-table-wrapper {
    padding: 0;
  }
  .ranking-table {
    width: 100%;
    font-size: 1rem;
    border-radius: 0;
    .header-row th {
      font-size: 1rem;
      border-radius: 0;
    }
    .row td {
      border-radius: 0;
    }
  }
  .user-info {
    flex-direction: column;
    img {
      margin: 0.5em 0 0 0;
    }
  }
}