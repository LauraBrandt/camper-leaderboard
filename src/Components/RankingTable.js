import React from 'react';
import TableRow from './TableRow';
import SortingButton from './SortingButton';
import users from '../data';

class RankingTable extends React.Component {
  constructor() {
    super();
    this.state = {
      users: [],
      inactive: 'recent up'
    }

    this.handleSort = this.handleSort.bind(this);
    this.renderRows = this.renderRows.bind(this);
  }

  componentDidMount() {
    this.setState({ users });
  }

  handleSort(e) {
    const type = e.currentTarget.classList[0];
    const direction = e.currentTarget.classList[1];
    const sign = direction === 'up' ? -1 : 1;
    
    const users = this.state.users;
    users.sort((a, b) => sign * (a[type] - b[type]));
    this.setState({ users, inactive: `${type} ${direction}` })
  }

  renderRows() {
    return this.state.users.map((user, i) => 
      <TableRow user={user} index={i} key={user.username} />
    )
  }

  render() {
    return (
      <div className="ranking-table-wrapper">
        <table className="ranking-table">
          <thead>
            <tr className="row header-row">
              <th>#</th>
              <th>Camper</th> 
              <th>
                <div className="points">
                  <span className="points-label">Points in last 30 days</span>
                  <div className="points-sorters">
                    <SortingButton type="recent" direction="up" inactive={this.state.inactive === 'recent up'} handleClick={this.handleSort}/>
                    <SortingButton type="recent" direction="down" inactive={this.state.inactive === 'recent down'} handleClick={this.handleSort}/>
                  </div>
                </div>
              </th>
              <th>
                <div className="points">
                  <span>All time points</span>
                  <div className="points-sorters">
                    <SortingButton type="alltime" direction="up" inactive={this.state.inactive === 'alltime up'} handleClick={this.handleSort}/>
                    <SortingButton type="alltime" direction="down" inactive={this.state.inactive === 'alltime down'} handleClick={this.handleSort}/>
                  </div>
                </div>
              </th>
            </tr>
          </thead>
          <tbody>
            {this.state.users && this.renderRows()}
          </tbody>
        </table>
      </div>
    );
  }
}

export default RankingTable;
