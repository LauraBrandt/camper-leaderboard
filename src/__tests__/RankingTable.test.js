import React from 'react';
import {shallow, mount} from 'enzyme';
import renderer from 'react-test-renderer';
import RankingTable from '../Components/RankingTable';
import TableRow from '../Components/TableRow';
import SortingButton from '../Components/SortingButton';

const fetch = require('node-fetch');
const users = [
  {username: "a", recent: 5, alltime: 10, img: "#"},
  {username: "b", recent: 4, alltime: 8, img: "#"},
  {username: "c", recent: 3, alltime: 6, img: "#"},
]

beforeEach(() => {
  fetch.resetMocks();
  fetch.mockResponse(JSON.stringify( users ));
});

describe('RankingTable', () => {
  it('renders and matches snapshot', (done) => {
    const component = renderer.create(<RankingTable />);

    setImmediate(() => {
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
      done();
    }, 0);
  });
  
  describe('content', () => {
    let component;
    beforeEach(() => {
      component = shallow(<RankingTable />);
    });

    it('has a table', () => {
      expect(component.find('table')).toHaveLength(1);
    });

    it('has 4 columns', () => {
      expect(component.find('table th')).toHaveLength(4);
    });

    it('has 4 sorting buttons', () => {
      expect(component.find(SortingButton)).toHaveLength(4);
    });

    it('contains the same number of TableRow components as users', () => {
      const fakeUsers = [
        {username: "", recent: 0, alltime: 0, img: ""}, 
        {username: "", recent: 0, alltime: 0, img: ""}, 
        {username: "", recent: 0, alltime: 0, img: ""}];
      component.setState({ users: fakeUsers });

      const numTableRows = component.find(TableRow).length;
      const numUsers = component.state('users').length;
      expect(numTableRows).toEqual(numUsers);
    });
  });

  describe('child components', () => {
    it('passes correct props to TableRow', () => {
      const component = shallow(<RankingTable />);
      const fakeUsers = [{username: "testuser", recent: 5, alltime: 10, img: "www.example.com"}];
      component.setState({ users: fakeUsers });

      const tableRow = component.find(TableRow).at(0);
      expect(tableRow.prop('index')).toBe(0);
      expect(tableRow.prop('user')).toEqual(fakeUsers[0]);
    });

    it('passes the correct props to the SortingButtons', () => {
      const component = shallow(<RankingTable />);

      const buttons = component.find(SortingButton);

      expect(buttons.at(0).prop('type')).toEqual('recent');
      expect(buttons.at(0).prop('direction')).toEqual('up');
      expect(buttons.at(0).prop('inactive')).toEqual(true);
      expect(buttons.at(0).prop('handleClick')).toEqual(component.instance().handleSort);
      
      expect(buttons.at(1).prop('type')).toEqual('recent');
      expect(buttons.at(1).prop('direction')).toEqual('down');
      expect(buttons.at(1).prop('inactive')).toBeFalsy();
      expect(buttons.at(1).prop('handleClick')).toEqual(component.instance().handleSort);
      
      expect(buttons.at(2).prop('type')).toEqual('alltime');
      expect(buttons.at(2).prop('direction')).toEqual('up');
      expect(buttons.at(2).prop('inactive')).toBeFalsy();
      expect(buttons.at(2).prop('handleClick')).toEqual(component.instance().handleSort);
      
      expect(buttons.at(3).prop('type')).toEqual('alltime');
      expect(buttons.at(3).prop('direction')).toEqual('down');
      expect(buttons.at(3).prop('inactive')).toBeFalsy();
      expect(buttons.at(3).prop('handleClick')).toEqual(component.instance().handleSort);
    });
  });

  describe('interactions', () => {
    let component;
    const state = {
      users: [
        {username: "a", recent: 1, alltime: 2},
        {username: "b", recent: 3, alltime: 1},
        {username: "c", recent: 2, alltime: 3}
      ],
      inactive: 'recent up' 
    }

    describe('handleSort', () => {
      beforeEach((done) => {
        component = shallow(<RankingTable />);
        setImmediate(() => {
          component.setState( state );
          done();
        }, 0);
      });

      it("sorts 'recent up' and sets new state.inactive correctly", () => {
        component.setState({ inactive: 'recent down' });
        const event = {
          currentTarget: {
            classList: ['recent', 'up']
          }
        }

        component.instance().handleSort(event);
        expect(component.state('users')[0].recent).toBe(3);
        expect(component.state('users')[2].recent).toBe(1);
        expect(component.state('inactive')).toBe('recent up');
      });

      it("sorts 'recent down' and sets new state.inactive correctly", () => {
        const event = {
          currentTarget: {
            classList: ['recent', 'down']
          }
        }

        component.instance().handleSort(event);
        expect(component.state('users')[0].recent).toBe(1);
        expect(component.state('users')[2].recent).toBe(3);
        expect(component.state('inactive')).toBe('recent down');
      });

      it("sorts 'alltime down' and sets new state.inactive correctly", () => {
        const event = {
          currentTarget: {
            classList: ['alltime', 'down']
          }
        }

        component.instance().handleSort(event);
        expect(component.state('users')[0].alltime).toBe(1);
        expect(component.state('users')[2].alltime).toBe(3);
        expect(component.state('inactive')).toBe('alltime down');
      });

      it("sorts 'alltime up' and sets new state.inactive correctly", () => {
        const event = {
          currentTarget: {
            classList: ['alltime', 'up']
          }
        }

        component.instance().handleSort(event);
        expect(component.state('users')[0].alltime).toBe(3);
        expect(component.state('users')[2].alltime).toBe(1);
        expect(component.state('inactive')).toBe('alltime up');
      });
    });

    describe('handleSort integration', () => {
      beforeEach((done) => {
        component = mount(<RankingTable />);
        setImmediate(() => {
          component.setState( state );
          done();
        }, 0);
      });

      it('calls handleSort correctly on buttonClick', () => {
        component.instance().handleSort = jest.fn();
        component.instance().forceUpdate();

        expect(component.instance().handleSort).not.toBeCalled();
        component.find('.recent.down').simulate('click');
        expect(component.instance().handleSort).toBeCalled();
        
        const argument = component.instance().handleSort.mock.calls[0][0]
        expect(argument.target.classList[0]).toBe('recent');
        expect(argument.target.classList[1]).toBe('down');

        component.instance().handleSort.mockRestore();
      });

      it('executes handleSort correctly on button click', () => {
        component.find('.recent.down').simulate('click');

        expect(component.state('users')[0].recent).toBe(1);
        expect(component.state('users')[2].recent).toBe(3);
        expect(component.state('inactive')).toBe('recent down');
      });
    });
  });

  describe('message', () => {
    it('sets initial message to "Loading..." and renders it', () => {
      const component = shallow(<RankingTable />);
      expect(component.state('message')).toBe("Loading . . .");
      expect(component.find('.message').text()).toContain("Loading . . .");
    });

    it("changes message to empty string on successful api call and doesn't render it", () => {
      const component = shallow(<RankingTable />);
      expect.assertions(2);
      return component.instance().apiCall()
        .then(() => {
          expect(component.state('message')).toBe('');
          component.update();
          expect(component.find('.message')).toHaveLength(0);
        })
    });

    it('sets error message on failed api call and renders it', () => {
      fetch.mockReject(new Error('error'));
      expect.assertions(2);
      const component = shallow(<RankingTable />);

      return component.instance().apiCall()
        .then(() => {
          expect(component.state('message')).toContain('Sorry');
          component.update();
          expect(component.find('.message').text()).toContain('Sorry');
        })
    });
  });

  describe('api call', () => {
    it('fetch called with correct url', () => {
      const component = shallow(<RankingTable />);
      expect(fetch).toBeCalled();
      const url = 'https://fcctop100.herokuapp.com/api/fccusers/top/recent';
      expect(fetch).toBeCalledWith(url);
    });

    it('sets state.users correctly after successful api call and renders them', () => {
      const component = shallow(<RankingTable />);
      expect.assertions(3);
      return component.instance().apiCall()
        .then(() => {
          expect(component.state('users')).toEqual(users);
          component.update();
          expect(component.find(TableRow)).toHaveLength(users.length);
          const firstUser = component.find(TableRow).at(0).prop('user');
          expect(firstUser).toEqual(users[0]);
        })
    });
  });
});