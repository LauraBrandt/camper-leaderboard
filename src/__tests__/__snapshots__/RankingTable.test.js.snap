// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RankingTable renders and matches snapshot 1`] = `
<div
  className="ranking-table-wrapper"
>
  <table
    className="ranking-table"
  >
    <thead>
      <tr
        className="row header-row"
      >
        <th>
          #
        </th>
        <th>
          Camper
        </th>
        <th>
          <div
            className="points"
          >
            <span>
              Points in last 30 days
            </span>
            <div
              className="points-sorters"
            >
              <div>
                <button
                  className="recent up inactive"
                  onClick={null}
                  title="Sort descending"
                >
                  <i
                    className="fas fa-caret-up"
                  />
                </button>
              </div>
              <div>
                <button
                  className="recent down"
                  onClick={[Function]}
                  title="Sort ascending"
                >
                  <i
                    className="fas fa-caret-down"
                  />
                </button>
              </div>
            </div>
          </div>
        </th>
        <th>
          <div
            className="points"
          >
            <span>
              All time points
            </span>
            <div
              className="points-sorters"
            >
              <div>
                <button
                  className="alltime up"
                  onClick={[Function]}
                  title="Sort descending"
                >
                  <i
                    className="fas fa-caret-up"
                  />
                </button>
              </div>
              <div>
                <button
                  className="alltime down"
                  onClick={[Function]}
                  title="Sort ascending"
                >
                  <i
                    className="fas fa-caret-down"
                  />
                </button>
              </div>
            </div>
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="row"
      >
        <td>
          1
        </td>
        <td>
          <a
            className="user-info"
            href="https://www.freecodecamp.org/a"
          >
            <img
              alt="a avatar"
              src="#"
            />
            <p>
              a
            </p>
          </a>
        </td>
        <td>
          5
        </td>
        <td>
          10
        </td>
      </tr>
      <tr
        className="row"
      >
        <td>
          2
        </td>
        <td>
          <a
            className="user-info"
            href="https://www.freecodecamp.org/b"
          >
            <img
              alt="b avatar"
              src="#"
            />
            <p>
              b
            </p>
          </a>
        </td>
        <td>
          4
        </td>
        <td>
          8
        </td>
      </tr>
      <tr
        className="row"
      >
        <td>
          3
        </td>
        <td>
          <a
            className="user-info"
            href="https://www.freecodecamp.org/c"
          >
            <img
              alt="c avatar"
              src="#"
            />
            <p>
              c
            </p>
          </a>
        </td>
        <td>
          3
        </td>
        <td>
          6
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
